### PR TITLE
[SPARK-19435][SQL] Type coercion between ArrayTypes

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR proposes to support type coercion between `ArrayType`s where the element types are compatible.

**Before**

```
Seq(Array(1)).toDF("a").selectExpr("greatest(a, array(1D))")
org.apache.spark.sql.AnalysisException: cannot resolve 'greatest(`a`, array(1.0D))' due to data type mismatch: The expressions should all have the same type, got GREATEST(array<int>, array<double>).; line 1 pos 0;

Seq(Array(1)).toDF("a").selectExpr("least(a, array(1D))")
org.apache.spark.sql.AnalysisException: cannot resolve 'least(`a`, array(1.0D))' due to data type mismatch: The expressions should all have the same type, got LEAST(array<int>, array<double>).; line 1 pos 0;

sql("SELECT * FROM values (array(0)), (array(1D)) as data(a)")
org.apache.spark.sql.AnalysisException: incompatible types found in column a for inline table; line 1 pos 14

Seq(Array(1)).toDF("a").union(Seq(Array(1D)).toDF("b"))
org.apache.spark.sql.AnalysisException: Union can only be performed on tables with the compatible column types. ArrayType(DoubleType,false) <> ArrayType(IntegerType,false) at the first column of the second table;;

sql("SELECT IF(1=1, array(1), array(1D))")
org.apache.spark.sql.AnalysisException: cannot resolve '(IF((1 = 1), array(1), array(1.0D)))' due to data type mismatch: differing types in '(IF((1 = 1), array(1), array(1.0D)))' (array<int> and array<double>).; line 1 pos 7;
```

**After**

```scala
Seq(Array(1)).toDF("a").selectExpr("greatest(a, array(1D))")
res5: org.apache.spark.sql.DataFrame = [greatest(a, array(1.0)): array<double>]

Seq(Array(1)).toDF("a").selectExpr("least(a, array(1D))")
res6: org.apache.spark.sql.DataFrame = [least(a, array(1.0)): array<double>]

sql("SELECT * FROM values (array(0)), (array(1D)) as data(a)")
res8: org.apache.spark.sql.DataFrame = [a: array<double>]

Seq(Array(1)).toDF("a").union(Seq(Array(1D)).toDF("b"))
res10: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [a: array<double>]

sql("SELECT IF(1=1, array(1), array(1D))")
res15: org.apache.spark.sql.DataFrame = [(IF((1 = 1), array(1), array(1.0))): array<double>]
```

## How was this patch tested?

Unit tests in `TypeCoercion` and Jenkins tests and

building with scala 2.10

```scala
./dev/change-scala-version.sh 2.10
./build/mvn -Pyarn -Phadoop-2.4 -Dscala-2.10 -DskipTests clean package
```

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 1]
```diff
@@ -101,24 +101,13 @@ object TypeCoercion {
...
-      case _ => None
-    })
-  }
-
   /**
-   * Find the tightest common type of a set of types by continuously applying
-   * `findTightestCommonTypeOfTwo` on these types.
+   * Promotes all the way to StringType.
    */
-  private def findTightestCommonType(types: Seq[DataType]): Option[DataType] = {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: `findTightestCommonType` was removed as it seems used nowhere.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala [hunk 2]
```diff
@@ -53,7 +53,8 @@ class TypeCoercionSuite extends PlanTest {
...
   // | CalendarIntervalType | X        | X         | X           | X        | X          | X         | X          | X          | X           | X          | X        | X             | X          | X        | X           | X        | CalendarIntervalType | X                   | X           | X            |
   // +----------------------+----------+-----------+-------------+----------+------------+-----------+------------+------------+-------------+------------+----------+---------------+------------+----------+-------------+----------+----------------------+---------------------+-------------+--------------+
-  // Note: ArrayType*, MapType*, StructType* are castable only when the internal child types also match; otherwise, not castable
+  // Note: MapType*, StructType* are castable only when the internal child types also match; otherwise, not castable.
+  // Note: ArrayType* is castable when the element type is castable according to the table.
```
Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala]: It seems we now support implicit cast of `ArrayType` via https://issues.apache.org/jira/browse/SPARK-18624.

**Comment by HyukjinKwon** [comment]
cc @hvanhovell, could you maybe take a look please?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: It becomes harder for reviewers to read this PR. Could you submit a separate PR for code cleaning? Thanks!

Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Sure, I submitted for this in https://github.com/apache/spark/pull/16786.

**Comment by HyukjinKwon** [comment]
cc @cloud-fan, WDYT?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 3]
```diff
@@ -101,13 +101,13 @@ object TypeCoercion {
...
-  def findTightestCommonTypeToString(left: DataType, right: DataType): Option[DataType] = {
-    findTightestCommonType(left, right).orElse((left, right) match {
-      case (StringType, t2: AtomicType) if t2 != BinaryType && t2 != BooleanType => Some(StringType)
-      case (t1: AtomicType, StringType) if t1 != BinaryType && t1 != BooleanType => Some(StringType)
-      case _ => None
-    })
+  /**
+   * Promotes all the way to StringType.
+   */
+  private def stringPromotion: (DataType, DataType) => Option[DataType] = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: this becomes a method that returns a function, is it what we need?

**Comment by cloud-fan** [comment]
@HyukjinKwon are you sure this is a common behavior in databases?

**Comment by HyukjinKwon** [comment]
Let me check other DBMSs and back.

**Comment by HyukjinKwon** [comment]
**Postgres**

```
postgres=# SELECT greatest(array[1], array[0.1]);
 greatest
----------
 {1}
(1 row)

postgres=# SELECT least(array[1], array[0.1]);
 least
-------
 {0.1}
(1 row)
```

```
postgres=# SELECT * FROM (values (array[1]), (array[0.1])) as foo;
 column1
---------
 {1}
 {0.1}
(2 rows)
```

```
postgres=# SELECT array[1] UNION SELECT array[0.1];
 array
-------
 {0.1}
 {1}
(2 rows)
```

```
postgres=# SELECT CASE WHEN TRUE THEN array[0.1] ELSE array[1] END;
 array
-------
 {0.1}
(1 row)
```

(sorry, I could not find a proper way to test this with `IF`. So, I used `CASE`/`WHEN` in postgres).




**Hive** - not supporting this type coercion.

```
SELECT least(array(1), array(1D));
FAILED: SemanticException [Error 10014]: Line 1:7 Wrong arguments '1': least only takes primitive types, got array<int>
```
least/greatest: seems only supporting primitive types

```
SELECT inline(array(struct(array(0)), struct(array(1D))));
FAILED: SemanticException [Error 10016]: Line 1:38 Argument type mismatch '1D': Argument type "struct<col1:array<double>>" is different from preceding arguments. Previous type was "struct<col1:array<int>>"
```

```
SELECT array(1) UNION SELECT array(1D);
FAILED: SemanticException Schema of both sides of union should match: Column _c0 is of type array<int> on first table and type array<double> on second table. Cannot tell the position of null AST.
```

```
SELECT IF(1=1, array(1), array(1D));
FAILED: SemanticException [Error 10016]: Line 1:25 Argument type mismatch '1D': The second and the third arguments of function IF should have the same type, but they are different: "array<int>" and "array<double>"
```

**MySQL**

Seems not supporting arrays

**Comment by HyukjinKwon** [comment]
@cloud-fan, To cut this short, it seems Postgres supports this whereas Hive does not.

It seems we now support implicit cast between`ArrayType`s via [SPARK-18624](https://issues.apache.org/jira/browse/SPARK-18624), for example as below:

```scala
sql("SELECT percentile_approx(10.0, array('1', '1', '1'), 100)").show()
```

Wouldn't it be more reasonable to allow this case?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 3 above)
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Oh, I will fix it.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala [hunk 4]
```diff
@@ -379,6 +380,75 @@ class TypeCoercionSuite extends PlanTest {
     widenTest(ArrayType(IntegerType), StructType(Seq()), None)
   }
 
+  test("wider common type for decimal and array") {
+    def widenTest(t1: DataType, t2: DataType, tightestCommon: Option[DataType]) {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala]: `tightestCommon` -> `expected`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala [hunk 5]
```diff
@@ -379,6 +380,75 @@ class TypeCoercionSuite extends PlanTest {
...
+      assert(found == tightestCommon,
+        s"Expected $tightestCommon as wider common type for $t2 and $t1, found $found")
+    }
+
+    // Decimal
+    widenTest(DecimalType(2, 1), DecimalType(3, 2), Some(DecimalType(3, 2)))
+    widenTest(DecimalType(2, 1), DoubleType, Some(DoubleType))
+    widenTest(DecimalType(2, 1), IntegerType, Some(DecimalType(11, 1)))
+    widenTest(DoubleType, DecimalType(2, 1), Some(DoubleType))
+    widenTest(IntegerType, DecimalType(2, 1), Some(DecimalType(11, 1)))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala]: `widenTest` already tests symmetric

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala [hunk 6]
```diff
@@ -379,6 +380,75 @@ class TypeCoercionSuite extends PlanTest {
...
+    widenTest(LongType, DecimalType(2, 1), Some(DecimalType(21, 1)))
+
+    // ArrayType
+    widenTest(ArrayType(NullType), ArrayType(ByteType), Some(ArrayType(ByteType)))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala]: These tests are too similar. we just need one test case for numeric, one for string, one for nested array.

**Comment by HyukjinKwon** [comment]
@cloud-fan, I just addressed your comments and test a build with Scala 2.10.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala [hunk 7]
```diff
@@ -379,6 +386,67 @@ class TypeCoercionSuite extends PlanTest {
...
+
+    // Decimal
+    widenTestWithStringPromotion(
+      DecimalType(2, 1), DecimalType(3, 2), Some(DecimalType(3, 2)))
+    widenTestWithStringPromotion(
+      DecimalType(2, 1), DoubleType, Some(DoubleType))
+    widenTestWithStringPromotion(
+      DecimalType(2, 1), IntegerType, Some(DecimalType(11, 1)))
+    widenTestWithStringPromotion(
+      DoubleType, DecimalType(2, 1), Some(DoubleType))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala]: nit: how about we always put the decimal type in left? then we won't mistakenly add symmetric test

Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala]: can we add a test for string? (not array(string))

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala [hunk 8]
```diff
@@ -379,6 +386,67 @@ class TypeCoercionSuite extends PlanTest {
...
+      ArrayType(TimestampType, containsNull = false),
+      ArrayType(StringType, containsNull = true),
+      Some(ArrayType(StringType, containsNull = true)))
+    widenTestWithStringPromotion(
+      ArrayType(ArrayType(IntegerType), containsNull = false),
+      ArrayType(ArrayType(LongType), containsNull = false),
+      Some(ArrayType(ArrayType(LongType), containsNull = false)))
+
+    // Without string promotion
+    widenTestWithoutStringPromotion(IntegerType, StringType, None)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala]: I think we can just test int or long, not both

**Comment by cloud-fan** [comment]
LGTM except several minor tests about tests, thanks for working on it!

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala (hunk 8 above)
Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala]: `LongType` test was removed.

**Comment by HyukjinKwon** [comment]
Thanks @cloud-fan for your detailed review. I will keep in mind those comments.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 9]
```diff
@@ -116,48 +114,66 @@ object TypeCoercion {
...
-    case (t: IntegralType, d: DecimalType) =>
-      Some(DecimalPrecision.widerDecimalType(DecimalType.forType(t), d))
-    case (d: DecimalType, t: IntegralType) =>
-      Some(DecimalPrecision.widerDecimalType(DecimalType.forType(t), d))
-    case (_: FractionalType, _: DecimalType) | (_: DecimalType, _: FractionalType) =>
-      Some(DoubleType)
-    case _ =>
-      findTightestCommonTypeToString(t1, t2)
+  def findWiderTypeForTwo(t1: DataType, t2: DataType): Option[DataType] = {
+    findTightestCommonType(t1, t2)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Previously, `findWiderTypeForDecimal ` is before `findTightestCommonTypeToString `.  Thus, the results could be different. cc @cloud-fan 

You changed the order. I am not sure whether this should be documented in the release note.

**Comment by gatorsmile** [comment]
I think we need to separate the changes from the support of `Type coercion between ArrayTypes`? Could you submit another PR at first? We might need extra test cases for this change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 9 above)
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Yes, it is true that the type dispatch order was changed but `findTightestCommonType` does not take care of `DecimalType` therefore the results would be the same.

**Comment by HyukjinKwon** [comment]
Do you mean two PRs for cleaning up the logics here and the support of array type coercion?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 9 above)
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: @cloud-fan refactored this logic recently and I believe he didn't missed this part.

**Comment by gatorsmile** [comment]
Yeah, the first PR is for refactoring and cleaning up `findWiderTypeForTwo`. We need to add the test cases for the behavior changes. We might also need to document this in the release note, because it changes the output types.

The second one is for `Type coercion between ArrayTypes`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 9 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: The result is the same? 

See the cases in `findTightestCommonType`: https://github.com/HyukjinKwon/spark/blob/510a0eee43030abbf37ef922684e6165d6f1e1c8/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L87-L90

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: The original `findTightestCommonTypeToString` does not handle `DecimalType `.  However, this PR is calling the `findTightestCommonType ` at first.

Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Aha, thank you for correcting me. I overlooked but the result should be still the same, shouldn't it?

- `DecimalType.isWiderThan`

```
(p1 - s1) >= (p2 - s2) && s1 >= s2
```

- DecimalPrecision.widerDecimalType

```
 max(s1, s2) + max(p1-s1, p2-s2), max(s1, s2)
```

If both are different, we were already applying different type coercion rules between `findWiderTypeWithoutStringPromotion` and `findWiderTypeForTwo`, I guess we should match them with the same given https://github.com/apache/spark/pull/14439 ?

**Comment by HyukjinKwon** [comment]
I see what you mean. The code paths are now different. Let me try to investigate it and split them.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 9 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: We should make them consistent. That is why I think it is right to make the change, even if it causes the behavior changes.

**Comment by HyukjinKwon** [comment]
@gatorsmile, Can we make this merged and then add test cases for them separately? It seems the results are the same. I ran two tests as below:

```scala
val integralTypes =
  IndexedSeq(
    ByteType,
    ShortType,
    IntegerType,
    LongType)

val decimals = (-38 to 38).flatMap { p =>
  (-38 to 38).flatMap(s => allCatch opt DecimalType(p, s))
}

assert(decimals.nonEmpty)
integralTypes.foreach { it =>
  test(s"$it test") {
    decimals.foreach { d =>

      // From TypeCoercion.findWiderTypeForTwo
      val maybeType1 = (d, it) match {
        case (d: DecimalType, t: IntegralType) =>
          Some(DecimalPrecision.widerDecimalType(DecimalType.forType(t), d))
        case _ => None
      }

      // From TypeCoercion.findTightestCommonType
      val maybeType2 = (d, it) match {
        case (t1: DecimalType, t2: IntegralType) if t1.isWiderThan(t2) =>
          Some(t1)
        case _ => None
      }

      if (maybeType2.isDefined) {
        val t1 = maybeType1.get
        val t2 = maybeType2.get
        assert(t1 == t2)
      }
    }
  }
}
```

```scala
val integralTypes =
  IndexedSeq(
    ByteType,
    ShortType,
    IntegerType,
    LongType)

val decimals = (-38 to 38).flatMap { p =>
  (-38 to 38).flatMap(s => allCatch opt DecimalType(p, s))
}
  
assert(decimals.nonEmpty)
  
integralTypes.foreach { it =>
  test(s"$it test") {
    val widenDecimals = decimals.flatMap { d =>
      // From TypeCoercion.findWiderTypeForTwo
      (d, it) match {
        case (d: DecimalType, t: IntegralType) =>
          Some(DecimalPrecision.widerDecimalType(DecimalType.forType(t), d))
        case _ => None
      }
    }.toSet

    val tightDecimals = decimals.flatMap { d =>
      // From TypeCoercion.findTightestCommonType
      (d, it) match {
        case (t1: DecimalType, t2: IntegralType) if t1.isWiderThan(t2) =>
          Some(t1)
        case _ => None
      }
    }.toSet

    assert(widenDecimals.nonEmpty)
    assert(tightDecimals.nonEmpty)
    assert(tightDecimals.subsetOf(widenDecimals))
  }
}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 9 above)
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: I see. Thank you for catching it.

**Comment by HyukjinKwon** [comment]
(I just rebased and added `private[analysis]` for consistency)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 10]
```diff
@@ -117,49 +115,67 @@ object TypeCoercion {
...
-        Some(DecimalPrecision.widerDecimalType(DecimalType.forType(t), d))
-      case (d: DecimalType, t: IntegralType) =>
-        Some(DecimalPrecision.widerDecimalType(DecimalType.forType(t), d))
-      case (_: FractionalType, _: DecimalType) | (_: DecimalType, _: FractionalType) =>
-        Some(DoubleType)
-      case _ =>
-        findTightestCommonTypeToString(t1, t2)
-    }
+    findTightestCommonType(t1, t2)
+      .orElse(findWiderTypeForDecimal(t1, t2))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: yea we changed the order, but looks like it won't change the result. `findWiderTypeForDecimal` will always return a result for decimal type and numeric type, and if `findTightestCommonType` can return a result, `findWiderTypeForDecimal` will return the same result. So it doesn't matter if we run `findTightestCommonType` before or after it.

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Yes. Integer will be promoted to a wider Decimal anyway.

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Comment by HyukjinKwon** [comment]
Thank you @gatorsmile
